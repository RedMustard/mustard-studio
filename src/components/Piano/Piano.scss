@import '../../style/variables.scss';
@import './BlackKey/BlackKey.scss';
@import './WhiteKey/WhiteKey.scss';

.piano {
    display: table-row;
    height: $pianoHeight;
    position: relative;

    & > * {
        user-select: none;
    }
}
