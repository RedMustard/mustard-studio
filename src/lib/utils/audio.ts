
export const getNoteFrequencyByKeyNumber = (keyNumber: number) => {
    const frequencyA4 = 440;
    // https://en.wikipedia.org/wiki/Piano_key_frequencies
    const getFrequencyEquation = () => frequencyA4 * 2 ** ((keyNumber - 49) / 12);
    const roundedFrequency = Math.round(getFrequencyEquation() * 10000) / 10000;
    return roundedFrequency;
};
