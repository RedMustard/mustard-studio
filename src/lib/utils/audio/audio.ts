
export const getFrequencyByKeyNumber = (keyNumber: number) => {
    const frequencyA4 = 440;
    // https://en.wikipedia.org/wiki/Piano_key_frequencies
    const getFrequencyEquation = () => frequencyA4 * 2 ** ((keyNumber - 49) / 12);
    const roundedFrequency = Math.round(getFrequencyEquation() * 10000) / 10000;
    return roundedFrequency;
};


export const getFrequencyByOctaveOffset = (octaveOffset: number, frequency: number) => {
    let calculatedFrequency;
    const offset = 2 * octaveOffset;

    if (offset < 0) {
        calculatedFrequency = frequency / offset;
    } else if (offset > 0) {
        calculatedFrequency = frequency * offset;
    } else {
        calculatedFrequency = frequency;
    }
    return calculatedFrequency;
};
