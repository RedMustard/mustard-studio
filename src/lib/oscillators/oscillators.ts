import { OSC_1_INITIAL_SETTINGS, OSC_2_INITIAL_SETTINGS, OSC_SUB_INITIAL_SETTINGS } from '../../constants';
import {
    Oscillator,
    OscillatorDetuneSetting,
    OscillatorId,
    OscillatorSettings,
} from '../../types/types';
import { getAudioContext } from '../audioContext/audioContext';
import { getFrequencyByOctaveOffset } from '../utils/audio/audio';
import { logger } from '../utils/logger/logger';

let oscillatorNodes: {
    [frequency: string]: {
        [T: string]: OscillatorNode[];
    }
} = {};

let oscillatorConfigs: {
    [key in OscillatorId]: Oscillator
} = {
    osc1: {
        panNode: undefined,
        gainNode: undefined,
        settings: {
            ...OSC_1_INITIAL_SETTINGS,
        },
    },
    osc2: {
        panNode: undefined,
        gainNode: undefined,
        settings: {
            ...OSC_2_INITIAL_SETTINGS,
        },
    },
    oscSub: {
        panNode: undefined,
        gainNode: undefined,
        settings: {
            ...OSC_SUB_INITIAL_SETTINGS,
        },
    },
};

const addOscillatorNode = (oscillatorNode: OscillatorNode, frequency: number, oscillatorId: OscillatorId) => {
    if (!oscillatorNodes[frequency]) {
        oscillatorNodes[frequency] = {
            [oscillatorId]: [oscillatorNode],
        };
    } else if (!oscillatorNodes[frequency][oscillatorId]) {
        oscillatorNodes[frequency][oscillatorId] = [oscillatorNode];
    } else {
        oscillatorNodes[frequency][oscillatorId].push(oscillatorNode);
    }
    logger.info('Added oscillator nodes for frequency', frequency);
};

export const setOscillatorGainNodeByOscillatorId = (oscillatorId: OscillatorId, gainNode: GainNode) => {
    // logger.info('Oscillator Gain Node set for id', oscillatorId);
    // logger.info('Oscillators are now: ', oscillatorConfigs);

    switch (oscillatorId) {
        case 'osc1':
            oscillatorConfigs.osc1.gainNode = gainNode;
            break;
        case 'osc2':
            oscillatorConfigs.osc2.gainNode = gainNode;
            break;
        case 'oscSub':
            oscillatorConfigs.oscSub.gainNode = gainNode;
            break;
        default:
            break;
    }
};

export const setOscillatorPanNodeByOscillatorId = (oscillatorId: OscillatorId, panNode: StereoPannerNode) => {
    // logger.info('Oscillator Pan Node set for id', oscillatorId);
    // logger.info('Oscillators are now: ', oscillatorConfigs);

    switch (oscillatorId) {
        case 'osc1':
            oscillatorConfigs.osc1.panNode = panNode;
            break;
        case 'osc2':
            oscillatorConfigs.osc2.panNode = panNode;
            break;
        case 'oscSub':
            oscillatorConfigs.oscSub.panNode = panNode;
            break;
        default:
            break;
    }
};

export const setOscillatorSettingsByOscillatorId = (
    oscillatorId: OscillatorId,
    oscillatorSettings: OscillatorSettings | OscillatorSettings & OscillatorDetuneSetting,
) => {
    // logger.info('Oscillator settings set for id', oscillatorId, 'with settings,', oscillatorSettings);
    // logger.info('Oscillators are now: ', oscillatorConfigs);
    switch (oscillatorId) {
        case 'osc1':
            oscillatorConfigs.osc1.settings = oscillatorSettings;
            break;
        case 'osc2':
            oscillatorConfigs.osc2.settings = oscillatorSettings;
            break;
        case 'oscSub':
            oscillatorConfigs.oscSub.settings = oscillatorSettings;
            break;
        default:
            break;
    }
};

export const getOscillatorsConfigs = () => oscillatorConfigs;

export const getOscillatorsNodes = () => oscillatorNodes;

export const resetOscillatorNodes = () => {
    oscillatorNodes = {};
};

export const resetOscillatorConfigs = () => {
    oscillatorConfigs = {
        osc1: {
            panNode: undefined,
            gainNode: undefined,
            settings: {
                ...OSC_1_INITIAL_SETTINGS,
            },
        },
        osc2: {
            panNode: undefined,
            gainNode: undefined,
            settings: {
                ...OSC_2_INITIAL_SETTINGS,
            },
        },
        oscSub: {
            panNode: undefined,
            gainNode: undefined,
            settings: {
                ...OSC_SUB_INITIAL_SETTINGS,
            },
        },
    };
};

export const startOscillators = (oscillatorFrequency: number) => {
    const audioContext = getAudioContext();

    (Object.keys(oscillatorConfigs) as OscillatorId[]).forEach((oscillatorId) => {
        const oscSettings = oscillatorConfigs[oscillatorId].settings;
        const oscGainNode = oscillatorConfigs[oscillatorId].gainNode;
        const oscEnabled = oscSettings.enabled;
        const oscType = oscSettings.type;
        const oscOctave = oscSettings.octave;
        const calculatedFrequency = getFrequencyByOctaveOffset(oscOctave, oscillatorFrequency);

        if (oscEnabled) {
            const oscillator = audioContext.createOscillator();
            addOscillatorNode(oscillator, oscillatorFrequency, oscillatorId);
            oscillator.type = oscType;
            oscillator.frequency.value = calculatedFrequency;

            if (oscSettings.hasOwnProperty('detune')) {
                oscillator.detune.value = (oscSettings as OscillatorDetuneSetting).detune;
            }
            oscillator.connect(oscGainNode);
            oscillator.start();
            logger.info(`Starting oscillator ${oscillatorId} with frequency`, calculatedFrequency);
        }
    });
};

export const stopOscillatorByFrequency = (frequency: number) => {
    logger.info('Stopping all oscillators for frequency', frequency);
    if (oscillatorNodes[frequency]) {
        (Object.keys(oscillatorNodes[frequency]) as OscillatorId[]).forEach((oscillatorId: OscillatorId) => {
            oscillatorNodes[frequency][oscillatorId].forEach((oscillator) => {
                oscillator.stop();
            });
        });
    }
};

export const stopOscillatorById = (oscillatorId: OscillatorId) => {
    logger.info('Stopping all oscillators for frequency', oscillatorId);
    (Object.keys(oscillatorNodes)).forEach((frequency) => {
        if (oscillatorNodes[frequency][oscillatorId]) {
            oscillatorNodes[frequency][oscillatorId].forEach((oscillator) => {
                oscillator.stop();
            });
        }
    });
};

export const stopOscillators = () => {
    logger.info('Stopping all oscillators');
    (Object.keys(oscillatorNodes)).forEach((frequency) => {
        (Object.keys(oscillatorNodes[frequency]) as OscillatorId[]).forEach((oscillatorId: OscillatorId) => {
            oscillatorNodes[frequency][oscillatorId].forEach((oscillator) => {
                oscillator.stop();
            });
        });
    });
    oscillatorNodes = {};
};
