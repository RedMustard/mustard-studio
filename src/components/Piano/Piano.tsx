import { h } from 'preact';
import { getFrequencyByKeyNumber } from '../../lib/utils/audio/audio';
import { BlackKey } from './BlackKey/BlackKey';
import { WhiteKey } from './WhiteKey/WhiteKey';
import { PianoKeyColor } from '../../types/types';
import { logger } from '../../lib/utils/logger/logger';
import { startOscillators, stopOscillators } from '../../lib/oscillators/oscillators';
import { PIANO_OCTAVE_KEY_COUNT } from '../../constants';


export const Piano = () => {
    const handleOnMouseDownAndOver = (keyNumber: number) => {
        const frequency = getFrequencyByKeyNumber(keyNumber);
        startOscillators(frequency);
    };

    const handleOnMouseUpAndLeave = () => {
        stopOscillators();
    };

    const whiteKey = (keyNumber: number) => (
        <WhiteKey
            onMouseDown={(e: MouseEvent) => {
                if (e.buttons === 1) {
                    handleOnMouseDownAndOver(keyNumber);
                    logger.info('Piano key pressed', keyNumber);
                }
            }}
            onMouseOver={(e: MouseEvent) => {
                if (e.buttons === 1) {
                    handleOnMouseDownAndOver(keyNumber);
                    logger.info('Piano key pressed', keyNumber);
                }
            }}
            onMouseUp={() => {
                handleOnMouseUpAndLeave();
            }}
            onMouseLeave={(e: MouseEvent) => {
                if (e.buttons === 1) {
                    handleOnMouseUpAndLeave();
                }
            }}
        />
    );

    const blackKey = (keyNumber: number) => (
        <BlackKey
            onMouseDown={(e: MouseEvent) => {
                if (e.buttons === 1) {
                    handleOnMouseDownAndOver(keyNumber);
                    logger.info('Piano key pressed', keyNumber);
                }
            }}
            onMouseOver={(e: MouseEvent) => {
                if (e.buttons === 1) {
                    handleOnMouseDownAndOver(keyNumber);
                    logger.info('Piano key pressed', keyNumber);
                }
            }}
            onMouseUp={() => {
                handleOnMouseUpAndLeave();
            }}
            onMouseLeave={(e: MouseEvent) => {
                if (e.buttons === 1) {
                    handleOnMouseUpAndLeave();
                }
            }}
        />
    );

    const getKeysByOctaveCount = (octaveCount: number, startingKey: number) => {
        const octaveKeyPattern: PianoKeyColor[] = ['white', 'black', 'white', 'black', 'white', 'white', 'black', 'white', 'black', 'white', 'black', 'white'];
        const keys: h.JSX.Element[] = [];

        for (let i = 0; i < octaveCount; i++) {
            octaveKeyPattern.forEach((key: PianoKeyColor, index: number) => {
                const keyNumber = index + (i * PIANO_OCTAVE_KEY_COUNT) + startingKey;
                if (key === 'white') {
                    keys.push(whiteKey(keyNumber));
                } else if (key === 'black') {
                    keys.push(blackKey(keyNumber));
                }
            });
        }
        return keys;
    };

    return (
        <div class="piano">
            {whiteKey(1)}
            {blackKey(2)}
            {whiteKey(3)}
            {getKeysByOctaveCount(7, 4)}
            {whiteKey(88)}
        </div>
    );
};
