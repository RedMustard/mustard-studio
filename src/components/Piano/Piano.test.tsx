import { h } from 'preact';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Piano } from './Piano';
import { startOscillatorsByFrequency, stopOscillators } from '../../lib/oscillators/oscillators';
import { getFrequencyByKeyNumber } from '../../lib/utils/audio/audio';

jest.mock('../../lib/studioService/studioServiceActions');
jest.mock('../../lib/oscillators/oscillators');
jest.mock('../../lib/utils/audio/audio');

describe('<Piano />', () => {
    afterEach(() => {
        jest.clearAllMocks();
    });

    it('renders with basic props', () => {
        const wrapper = shallow(
            <Piano />,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('gets note frequency and starts oscillators on white key mousedown', () => {
        const wrapper = mount(
            <Piano />,
        );
        wrapper.find('.piano-key.piano-key--white').first().simulate('mousedown', { buttons: 1 });
        expect(startOscillatorsByFrequency).toBeCalledTimes(1);
        expect(getFrequencyByKeyNumber).toBeCalledTimes(1);
    });

    it('gets note frequency and starts oscillators on white key mouseover after mousedown', () => {
        const wrapper = mount(
            <Piano />,
        );
        wrapper.find('.piano-key.piano-key--white').first().simulate('mousedown', { buttons: 1 });
        wrapper.find('.piano-key.piano-key--white').last().simulate('mouseover', { buttons: 1 });
        expect(startOscillatorsByFrequency).toBeCalledTimes(2);
        expect(getFrequencyByKeyNumber).toBeCalledTimes(2);
    });

    it('stops oscillators on white key mouseup after mousedown', () => {
        const wrapper = mount(
            <Piano />,
        );
        wrapper.find('.piano-key.piano-key--white').first().simulate('mousedown', { buttons: 1 });
        wrapper.find('.piano-key.piano-key--white').first().simulate('mouseup', { buttons: 1 });
        expect(stopOscillators).toBeCalledTimes(1);
    });

    it('stops oscillators on white key mouseleave after mousedown', () => {
        const wrapper = mount(
            <Piano />,
        );
        wrapper.find('.piano-key.piano-key--white').first().simulate('mousedown', { buttons: 1 });
        wrapper.find('.piano-key.piano-key--white').first().simulate('mouseleave', { buttons: 1 });
        expect(stopOscillators).toBeCalledTimes(1);
    });

    it('stops oscillators on white key mouseup after mousedown when button 1 is not pressed', () => {
        const wrapper = mount(
            <Piano />,
        );
        wrapper.find('.piano-key.piano-key--white').first().simulate('mousedown', { buttons: 2 });
        wrapper.find('.piano-key.piano-key--white').first().simulate('mouseup', { buttons: 2 });
        expect(stopOscillators).toBeCalledTimes(1);
    });

    it('does not get note frequency or start oscillators on mousedown on white key when button 1 is not pressed', () => {
        const wrapper = mount(
            <Piano />,
        );
        wrapper.find('.piano-key.piano-key--white').first().simulate('mousedown', { buttons: 2 });
        expect(startOscillatorsByFrequency).not.toBeCalled();
        expect(getFrequencyByKeyNumber).not.toBeCalled();
    });

    it('does not get note frequency or start oscillators on white key mouseover after mousedown when button 1 is not pressed', () => {
        const wrapper = mount(
            <Piano />,
        );
        wrapper.find('.piano-key.piano-key--white').first().simulate('mousedown', { buttons: 2 });
        wrapper.find('.piano-key.piano-key--white').last().simulate('mouseover', { buttons: 2 });
        expect(startOscillatorsByFrequency).not.toBeCalled();
        expect(getFrequencyByKeyNumber).not.toBeCalled();
    });

    it('does not stop oscillators on white key mouseleave after mousedown when button 1 is not pressed', () => {
        const wrapper = mount(
            <Piano />,
        );
        wrapper.find('.piano-key.piano-key--white').first().simulate('mousedown', { buttons: 2 });
        wrapper.find('.piano-key.piano-key--white').first().simulate('mouseleave', { buttons: 2 });
        expect(stopOscillators).not.toBeCalled();
    });

    it('gets note frequency and starts oscillators on black key mousedown', () => {
        const wrapper = mount(
            <Piano />,
        );
        wrapper.find('.piano-key.piano-key--black').first().simulate('mousedown', { buttons: 1 });
        expect(startOscillatorsByFrequency).toBeCalledTimes(1);
        expect(getFrequencyByKeyNumber).toBeCalledTimes(1);
    });

    it('gets note frequency and starts oscillators on black key mouseover after mousedown', () => {
        const wrapper = mount(
            <Piano />,
        );
        wrapper.find('.piano-key.piano-key--black').first().simulate('mousedown', { buttons: 1 });
        wrapper.find('.piano-key.piano-key--black').last().simulate('mouseover', { buttons: 1 });
        expect(startOscillatorsByFrequency).toBeCalledTimes(1 * 2);
        expect(getFrequencyByKeyNumber).toBeCalledTimes(2);
    });

    it('stops oscillators on black key mouseup after mousedown', () => {
        const wrapper = mount(
            <Piano />,
        );
        wrapper.find('.piano-key.piano-key--black').first().simulate('mousedown', { buttons: 1 });
        wrapper.find('.piano-key.piano-key--black').first().simulate('mouseup', { buttons: 1 });
        expect(stopOscillators).toBeCalledTimes(1);
    });

    it('stops oscillators on black key mouseleave after mousedown', () => {
        const wrapper = mount(
            <Piano />,
        );
        wrapper.find('.piano-key.piano-key--black').first().simulate('mousedown', { buttons: 1 });
        wrapper.find('.piano-key.piano-key--black').first().simulate('mouseleave', { buttons: 1 });
        expect(stopOscillators).toBeCalledTimes(1);
    });

    it('stops oscillators on black key mouseup after mousedown when button 1 is not pressed', () => {
        const wrapper = mount(
            <Piano />,
        );
        wrapper.find('.piano-key.piano-key--black').first().simulate('mousedown', { buttons: 2 });
        wrapper.find('.piano-key.piano-key--black').first().simulate('mouseup', { buttons: 2 });
        expect(stopOscillators).toBeCalledTimes(1);
    });

    it('does not get note frequency or start oscillators on mousedown on black key when button 1 is not pressed', () => {
        const wrapper = mount(
            <Piano />,
        );
        wrapper.find('.piano-key.piano-key--black').first().simulate('mousedown', { buttons: 2 });
        expect(startOscillatorsByFrequency).not.toBeCalled();
        expect(getFrequencyByKeyNumber).not.toBeCalled();
    });

    it('does not get note frequency or start oscillators on black key mouseover after mousedown when button 1 is not pressed', () => {
        const wrapper = mount(
            <Piano />,
        );
        wrapper.find('.piano-key.piano-key--black').first().simulate('mousedown', { buttons: 2 });
        wrapper.find('.piano-key.piano-key--black').last().simulate('mouseover', { buttons: 2 });
        expect(startOscillatorsByFrequency).not.toBeCalled();
        expect(getFrequencyByKeyNumber).not.toBeCalled();
    });

    it('does not stop oscillators on black key mouseleave after mousedown when button 1 is not pressed', () => {
        const wrapper = mount(
            <Piano />,
        );
        wrapper.find('.piano-key.piano-key--black').first().simulate('mousedown', { buttons: 2 });
        wrapper.find('.piano-key.piano-key--black').first().simulate('mouseleave', { buttons: 2 });
        expect(stopOscillators).not.toBeCalled();
    });
});
