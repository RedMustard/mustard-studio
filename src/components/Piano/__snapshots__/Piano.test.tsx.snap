// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Piano /> renders with basic props 1`] = `
<div
  className="piano"
>
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <BlackKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
  <WhiteKey
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
  />
</div>
`;
